pkg/controller/instance/instance_controller.go

```diff
@@ -141,12 +150,28 @@ func (r *Reconciler) Reconcile(request ctrl.Request) (ctrl.Result, error) {
 		return reconcile.Result{}, err // OV not found has to be retried because it can really have been created after Instance
 	}
 
-	// ---------- 2. First check if we should start execution of new plan ----------
+	// ---------- 2. Check if the object is being deleted, start cleanup plan ----------
+
+	var isDeleting bool
+
+	// a delete request is indicated by a non-zero 'metadata.deletionTimestamp',
+	// see https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#finalizers
+	if instance.ObjectMeta.DeletionTimestamp.IsZero() {
+		if _, hasCleanupPlan := ov.Spec.Plans[kudov1beta1.CleanupPlanName]; hasCleanupPlan {
+			maybeAddFinalizer(instance)
+		}
+	} else {
+		log.Printf("InstanceController: Instance %s/%s is being deleted", instance.Namespace, instance.Name)
+		isDeleting = true
```

Comment:
does it really make sense to introduce a field here? You might as well have a method on instance instead or compute this inside `GetPlanToBeExecuted ` since that's the only usage.

Reply:
Good point, will do this instead!